### PR TITLE
rrd improvements

### PR DESCRIPTION
**Pull request by mambelli**
Code improvements in RRD handling code, no specific bug fix:
- Improved path handling and directory creations
- creating also parent directories,
- eliminated nested `os.path.join` calls,
- consolidated tmp2final verifyHelper functions,
- eliminated global variable use,
- use `os.path.join` instead of string concatenation with `/`

Also improved docstrings

The exception improvements changes have been separated out in PR #330

### CODE REVIEW

File: factory/glideFactoryMonitorAggregator.py [hunk 1]
```diff
@@ -913,15 +827,18 @@ def aggregateRRDStats(log=logSupport.log):
         # assigns the data from every site to 'stats'
         stats = {}
         for entry in monitorAggregatorConfig.entries:
-            rrd_fname = os.path.join(os.path.join(monitorAggregatorConfig.monitor_dir, "entry_" + entry), rrd_site(rrd))
+            rrd_fname = os.path.join(monitorAggregatorConfig.monitor_dir, "entry_" + entry, rrd_site(rrd))
             try:
                 stats[entry] = xmlParse.xmlfile2dict(rrd_fname, always_singular_list={"timezone": {}})
+            except FileNotFoundError:
+                log.debug(
+                    "aggregateRRDStats %s exception: parse_xml, IOError, File not found (OK if first time)" % rrd_fname
```
Comment (namrathaurs) [inline factory/glideFactoryMonitorAggregator.py:835]: Could this also be changed to be an f-string?

File: factory/glideFactoryMonitorAggregator.py [hunk 2]
```diff
@@ -913,15 +827,18 @@ def aggregateRRDStats(log=logSupport.log):
...
+                )
             except OSError:
-                if os.path.exists(rrd_fname):
-                    log.debug("aggregateRRDStats %s exception: parse_xml, IOError" % rrd_fname)
-                else:
+                log.debug("aggregateRRDStats %s exception: parse_xml, IOError" % rrd_fname)
```
Comment (namrathaurs) [inline factory/glideFactoryMonitorAggregator.py:838]: same comment as above

File: factory/glideFactoryMonitorAggregator.py [hunk 3]
```diff
@@ -913,15 +827,18 @@ def aggregateRRDStats(log=logSupport.log):
...
+                if not os.path.exists(rrd_fname):
                     log.debug(
-                        "aggregateRRDStats %s exception: parse_xml, IOError, File not existing (OK if first time)"
+                        "aggregateRRDStats %s exception: parse_xml, IOError, File not found (OK if first time) - should have been FileNotFoundError"
                         % rrd_fname
```
Comment (namrathaurs) [inline factory/glideFactoryMonitorAggregator.py:842]: same comment as above

File: factory/glideFactoryMonitorAggregator.py [hunk 4]
```diff
@@ -280,11 +182,12 @@ def aggregateStatus(in_downtime):
     for entry in monitorAggregatorConfig.entries:
         # load entry status file
         status_fname = os.path.join(
-            os.path.join(monitorAggregatorConfig.monitor_dir, "entry_" + entry), monitorAggregatorConfig.status_relname
+            monitorAggregatorConfig.monitor_dir, "entry_" + entry, monitorAggregatorConfig.status_relname
```
Comment (namrathaurs) [inline factory/glideFactoryMonitorAggregator.py:185]: Would it be better to change `"entry_" + entry` to `f"entry_{entry}"`?

File: factory/glideFactoryMonitorAggregator.py [hunk 5]
```diff
@@ -280,11 +182,12 @@ def aggregateStatus(in_downtime):
...
         )
         # load entry completed data file
         completed_data_fname = os.path.join(
-            os.path.join(monitorAggregatorConfig.monitor_dir, "entry_" + entry),
+            monitorAggregatorConfig.monitor_dir,
+            "entry_" + entry,
```
Comment (namrathaurs) [inline factory/glideFactoryMonitorAggregator.py:190]: Would it be better to change `"entry_" + entry` to `f"entry_{entry}"`?

File: factory/glideFactoryMonitorAggregator.py [hunk 6]
```diff
@@ -501,11 +404,13 @@ def aggregateStatus(in_downtime):
...
 
-    glideFactoryMonitoring.monitoringConfig.write_rrd_multi("total/Status_Attributes", "GAUGE", updated, val_dict)
+    glideFactoryMonitoring.monitoringConfig.write_rrd_multi(
+        os.path.join("total", "Status_Attributes"), "GAUGE", updated, val_dict
+    )
 
     # Frontend total rrds across all factories
     for fe in list(status_fe["frontends"].keys()):
-        glideFactoryMonitoring.monitoringConfig.establish_dir("total/%s" % ("frontend_" + fe))
+        glideFactoryMonitoring.monitoringConfig.establish_dir(os.path.join("total", "frontend_" + fe))
```
Comment (namrathaurs) [inline factory/glideFactoryMonitorAggregator.py:413]: Would it be better to change `"frontend_" + fe` to `f"frontend_{fe}"`?

File: factory/glideFactoryMonitorAggregator.py [hunk 7]
```diff
@@ -520,7 +425,7 @@ def aggregateStatus(in_downtime):
                     a_el = int(tp_el[a])
                     val_dict[f"{tp_str}{a}"] = a_el
         glideFactoryMonitoring.monitoringConfig.write_rrd_multi(
-            "total/%s/Status_Attributes" % ("frontend_" + fe), "GAUGE", updated, val_dict
+            os.path.join("total", "frontend_" + fe, "Status_Attributes"), "GAUGE", updated, val_dict
```
Comment (namrathaurs) [inline factory/glideFactoryMonitorAggregator.py:428]: Would it be better to change `"frontend_" + fe` to `f"frontend_{fe}"`?

File: factory/glideFactoryMonitorAggregator.py [hunk 8]
```diff
@@ -547,13 +452,20 @@ def aggregateJobsSummary():
     for entry in monitorAggregatorConfig.entries:
         # load entry log summary file
         status_fname = os.path.join(
-            os.path.join(monitorAggregatorConfig.monitor_dir, "entry_" + entry),
+            monitorAggregatorConfig.monitor_dir,
+            "entry_" + entry,
```
Comment (namrathaurs) [inline factory/glideFactoryMonitorAggregator.py:456]: Would it be better to change `"entry_" + entry` to `f"entry_{entry}"`?

File: factory/glideFactoryMonitorAggregator.py [hunk 9]
```diff
@@ -631,13 +543,15 @@ def aggregateLogSummary():
...
```
Comment (namrathaurs) [inline factory/glideFactoryMonitorAggregator.py:547]: Would it be better to change `"entry_" + entry` to `f"entry_{entry}"`?

File: factory/glideFactoryMonitorAggregator.py [hunk 10]
```diff
@@ -913,15 +827,18 @@ def aggregateRRDStats(log=logSupport.log):
         # assigns the data from every site to 'stats'
         stats = {}
         for entry in monitorAggregatorConfig.entries:
-            rrd_fname = os.path.join(os.path.join(monitorAggregatorConfig.monitor_dir, "entry_" + entry), rrd_site(rrd))
+            rrd_fname = os.path.join(monitorAggregatorConfig.monitor_dir, "entry_" + entry, rrd_site(rrd))
```
Comment (namrathaurs) [inline factory/glideFactoryMonitorAggregator.py:830]: Would it be better to change `"entry_" + entry` to `f"entry_{entry}"`?

File: frontend/glideinFrontendMonitorAggregator.py [hunk 11]
```diff
@@ -276,7 +201,7 @@ def aggregateStatus():
     for group in monitorAggregatorConfig.groups:
         # load group status file
         status_fname = os.path.join(
-            os.path.join(monitorAggregatorConfig.monitor_dir, "group_" + group), monitorAggregatorConfig.status_relname
+            monitorAggregatorConfig.monitor_dir, "group_" + group, monitorAggregatorConfig.status_relname
```
Comment (namrathaurs) [inline frontend/glideinFrontendMonitorAggregator.py:204]: Will it be better to change `"group_" + group` to `f"group_{group}"`?

File: frontend/glideinFrontendMonitorAggregator.py [hunk 12]
```diff
@@ -457,15 +382,15 @@ def aggregateStatus():
     # Write rrds
 
     glideinFrontendMonitoring.monitoringConfig.establish_dir("total")
-    write_one_rrd("total/Status_Attributes", updated, global_total, 0)
+    write_one_rrd(os.path.join("total", "Status_Attributes"), updated, global_total, 0)
 
     for fact in list(global_fact_totals["factories"].keys()):
-        fe_dir = "total/factory_%s" % glideinFrontendMonitoring.sanitize(fact)
+        fe_dir = os.path.join("total", "factory_" + glideinFrontendMonitoring.sanitize(fact))
```
Comment (namrathaurs) [inline frontend/glideinFrontendMonitorAggregator.py:388]: Will it be better to change to `f"factory_{glideinFrontendMonitoring.sanitize(fact)}"`?

File: frontend/glideinFrontendMonitorAggregator.py [hunk 13]
```diff
@@ -457,15 +382,15 @@ def aggregateStatus():
...
         glideinFrontendMonitoring.monitoringConfig.establish_dir(fe_dir)
-        write_one_rrd("%s/Status_Attributes" % fe_dir, updated, global_fact_totals["factories"][fact], 1)
+        write_one_rrd(os.path.join(fe_dir, "Status_Attributes"), updated, global_fact_totals["factories"][fact], 1)
     for fact in list(global_fact_totals["states"].keys()):
-        fe_dir = "total/state_%s" % glideinFrontendMonitoring.sanitize(fact)
+        fe_dir = os.path.join("total", "state_" + glideinFrontendMonitoring.sanitize(fact))
```
Comment (namrathaurs) [inline frontend/glideinFrontendMonitorAggregator.py:392]: Will it be better to change to `f"state_{glideinFrontendMonitoring.sanitize(fact)}"`?

**Review by namrathaurs — CHANGES_REQUESTED** [review]
Almost all the suggestions are similar to each other; with only a change in the respective variables depending on what's being used. Please consider these comments if valid and applicable. If not valid, I'd be interested to know why they are not.

File: frontend/glideinFrontendMonitorAggregator.py (hunk 13 above)
Comment (mambelli) [reply frontend/glideinFrontendMonitorAggregator.py:392]: done

File: frontend/glideinFrontendMonitorAggregator.py (hunk 12 above)
Comment (mambelli) [reply frontend/glideinFrontendMonitorAggregator.py:388]: done

File: frontend/glideinFrontendMonitorAggregator.py (hunk 11 above)
Comment (mambelli) [reply frontend/glideinFrontendMonitorAggregator.py:204]: done

File: factory/glideFactoryMonitorAggregator.py (hunk 10 above)
Comment (mambelli) [reply factory/glideFactoryMonitorAggregator.py:830]: done

File: factory/glideFactoryMonitorAggregator.py (hunk 9 above)
Comment (mambelli) [reply factory/glideFactoryMonitorAggregator.py:547]: done

File: factory/glideFactoryMonitorAggregator.py (hunk 8 above)
Comment (mambelli) [reply factory/glideFactoryMonitorAggregator.py:456]: done

File: factory/glideFactoryMonitorAggregator.py (hunk 7 above)
Comment (mambelli) [reply factory/glideFactoryMonitorAggregator.py:428]: done

File: factory/glideFactoryMonitorAggregator.py (hunk 6 above)
Comment (mambelli) [reply factory/glideFactoryMonitorAggregator.py:413]: done

File: factory/glideFactoryMonitorAggregator.py (hunk 5 above)
Comment (mambelli) [reply factory/glideFactoryMonitorAggregator.py:190]: done

File: factory/glideFactoryMonitorAggregator.py (hunk 4 above)
Comment (mambelli) [reply factory/glideFactoryMonitorAggregator.py:185]: done

File: factory/glideFactoryMonitorAggregator.py [hunk 14]
```diff
@@ -89,154 +77,68 @@ def rrd_site(name):
...
         "Status_Attributes.rrd": status_dict,
         "Log_Completed.rrd": completed_dict,
         "Log_Completed_Stats.rrd": completed_stats_dict,
         "Log_Completed_WasteTime.rrd": completed_waste_dict,
         "Log_Counts.rrd": counts_dict,
     }
-
+    # check all the existing files
+    if not os.path.isdir(mon_dir):
+        print("WARNING: monitor/ directory does not exist, skipping rrd verification.")
```
[the remark this answers is not in the document]
Comment (mambelli) [reply factory/glideFactoryMonitorAggregator.py:135]: done

File: lib/rrdSupport.py [hunk 15]
```diff
@@ -672,3 +661,67 @@ def addDataStore(filenamein, filenameout, attrlist):
...
+        (out, tempfilename2) = tempfile.mkstemp()
+        (restored, restoredfilename) = tempfile.mkstemp()
+        os.close(out)
+        os.close(restored)
+        os.unlink(restoredfilename)
+        # Use exe version since dump, restore not available in rrdtool
+        dump_obj = rrdtool_exe()
+        outstr = dump_obj.dump(filename)
+        for line in outstr:
+            os.write(f, "%s\n" % line)
```
[the remark this answers is not in the document]
Comment (mambelli) [reply lib/rrdSupport.py:710]: done, and fixed also the encoding

File: lib/rrdSupport.py [hunk 16]
```diff
@@ -672,3 +661,67 @@ def addDataStore(filenamein, filenameout, attrlist):
...
+        fix_rrd(bool): if True, will attempt to add missing attrs
+        backup(bool): if not True skip the backup of original rrd
+
+    Returns:
+        bool: True if there were some problem with the RRD file, False if all OK
+
+    """
+    rrd_problems_found = False
+    if not os.path.exists(filename):
+        print("WARNING: %s missing, will be created on restart" % (filename))
```
[the remark this answers is not in the document]
Comment (mambelli) [reply lib/rrdSupport.py:686]: done

**Review by namrathaurs — APPROVED** [review]
Changes look good and can be merged!

**Pull request merged**